### ISSUE TITLE
Add codecov configuration to enforce 100% patch coverage

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8494]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8494|
|Type|enhancement|
|Created|2016-06-23 08:12:02Z|
|Branch|https://github.com/twisted/twisted/tree/8494-codecov-config|

I think that we need a custom codecov.io configuration file to make sure that patch get 100% coverage

AFAIK patch coverage is only required to be at 80% 

We should also update the dev docs to talk about codecov.io and make it official

<details><summary>Searchable metadata</summary>

```
trac-id__8494 8494
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__8494_codecov_config 8494-codecov-config
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1466669522776027 1466669522776027
changetime__1467219245173157 1467219245173157
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban commented|
|-|-|

I added a codecov config to make sure patch is 100%.


I have also enabled travis build for every branch so that we don't need a PR for each branch that we want to land :)

Please check that all is ok. Thanks!

----------------

In this PR https://github.com/twisted/twisted/pull/192 you can see that commit 4f15df8 has passed the patch coverage, even if it has coverage 97.29% of diff

------

From what I can tell codecov is using some heuristic technique to detect when all coverage are reported for a branch... so it might delay reporting the status

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @adiroiban|
|-|-|

Thanks for this, Adi.

The comment is inaccurate.  In fact codecov does have a way to ensure overall coverage increases: `coverage.status.project.default.target = auto`; have a look in the docs.  I believe this is actually the default behavior though, so we don't need to do much :).

The settings are fine though; please fix the comment and land.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

Please check this PR https://github.com/twisted/twisted/pull/259

the report was only sent from Travis-CI so it is expected to have decreased coverage.

The codecov/project is green but it reports "Absolute coverage decreased by -1.70%"

I don't make sense of the current documentation in codecov.io

```
Coverage Target

number; 0 <= N <= 100 | auto

auto: (default) Codecov will compare coverage against the parent commit to determin if coverage increased or decreased.

number: Codecov will set the status based on if coveage was greater then or equal to the target.
```

There is also this codecov.io but report, but I have no idea if this was fixed or not.... it talked about flagged repos https://github.com/codecov/support/issues/68

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

OK.

so the actual text is 

```
Absolute coverage decreased by -0.69% but relative coverage increased by +10.23% compared to bb0d1d6
         
```

I will update the comment and merge.
Thanks!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|GitHub set status to `closed`|
|-|-|

In changeset 747271739a61f624317b313a0d31b45dcf0cb714
```
#!CommitTicketReference repository="" revision="747271739a61f624317b313a0d31b45dcf0cb714"
Merge 8494-codecov-config: Add dedicated codecov configuration.


Author: adiroiban
Reviewer: glyph
Fixes: #8494
```
